Display previously viewed screen when switching back to Site, Notification, and Reader

crazytonyli opened the pull request:
## Issue

1. In iPad full-screen mode, open sidebar, select the Reader(*).
1. View a post's comments from the Reader.
1. Open sidebar, select Notifications.
1. Go back to the Reader.

I'd expect the content in step 2 is displayed. But the actual result is the root-level of the Reader is displayed.

(*) The same issue can happen on a site or the Notifications.

## Changes

This PR preserves the previously viewed content by keeping the relevant view controllers in memory and using them instead of creating new ones when switching back.

At the moment, other than sidebar, there are three things that can be displayed within the split view: site details, Notifications, and Reader. I have create dedicated types for those content, so that they are better organized in code than storing view controller instances directly as `SplitViewRootPresenter` properties.

Please note, when switching between sites, the previously viewed screen in the last site will still be lost. It'd be a simple change to preserve that too. I'm open for suggestions, but I don't feel like it's as important as preserving previously viewed screen in the Notifications and Reader.

## Regression Notes
1. Potential unintended areas of impact


2. What I did to test those areas of impact (or what existing automated tests I relied on)


3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding unit tests for my changes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

crazytonyli committed:
Display last screen when switching back to Site, Notification, and Reader

crazytonyli committed:
Remove Notifications and Reader upon low memory warning

crazytonyli committed:
Add an accidentally deleted line back

kean commented on WordPress/Classes/System/SplitViewRootPresenter.swift:
```diff
@@ -361,17 +389,50 @@ extension SplitViewRootPresenter: UISplitViewControllerDelegate {
-            splitVC.setViewController(navigationVC, for: .secondary)
+    func splitViewController(_ svc: UISplitViewController, willShow column: UISplitViewController.Column) {
+        if column == .primary, let selection = displayingContent?.selection, selection != sidebarViewModel.selection {
+            sidebarViewModel.selection = selection
```
I'm not fully sure what this change does, but the source of truth should always be `sidebarViewModel.selection`. The ViewModel shouldn't react to selection made by `UISplitViewController`.

If this code is removed, I suggest also removing `var selection: SidebarSelection` from `SplitViewDisplayable` as these screens don't really need to know about how the sidebar selects them.

crazytonyli replied:
> the source of truth should always be sidebarViewModel.selection. 

That'd be my preference too. However, I think it's a bit hard to guarantee that. For example, any place can call the `configure(selection:)` function instead of `sidebarViewModel.selection = ...`. They should not do that, but they can.

If we'd like to make sure the data and view in-sync, it's probably best to make it impossible to programmatically updating view without updating the data.

crazytonyli replied:
I'll update the PR to always use `sidebarViewModel.selection` to update UI. We can figure out the "make it impossible to programmatically updating view without updating the data" part later.

kean commented on WordPress/Classes/System/SplitViewRootPresenter.swift:
```diff
@@ -361,17 +389,50 @@ extension SplitViewRootPresenter: UISplitViewControllerDelegate {
+/// content, depending on what users choose from the sidebar.
+protocol SplitViewDisplayable: AnyObject {
+    var supplimentary: UINavigationController { get }
+    var secondary: UINavigationController? { get set }
```
`var secondary: UINavigationController?` should not be optional because it you have to display something in the `.secondary` tab, even if has to be a "Not Found" or "Something Went Wrong" or "No Selection" screen.

kean replied:
**Note**: I think it's OK to hardcode secondary to `UINavigationController` because that's the default `UISplitViewController` behavior. By default, `UISplitViewController` retains the same `UINavigationController` when you change the `.secondary` column creating a navigation stack. I disabled this behavior by making `.secondary` column a navigation controller. It's unlikely that it can ever be anything else and you can always hide the navigation bar completely if that's what you want.

kean commented on WordPress/Classes/System/SplitViewRootPresenter+Reader.swift:
```diff
@@ -0,0 +1,40 @@
+class ReaderSplitViewContent: SplitViewDisplayable {
+    let sidebar: ReaderSidebarViewController
+    let sidebarNavigationController: UINavigationController
+    var content: UINavigationController
```
If `secondary` becomes non-optional (my other comment), there is no need for a separate `content` . It can be a it confusing that it has both. In the same vein, `sidebarNavigationController` could also be merged with `supplimentary`, there appears to be no need for a separate property.

kean commented on WordPress/Classes/System/SplitViewRootPresenter+Reader.swift:
```diff
@@ -0,0 +1,40 @@
+
+    func displayed(in splitVC: UISplitViewController) {
+        if content.viewControllers.isEmpty {
+            sidebar.showInitialSelection()
```
There is a disconnect here because `ReaderSidebarViewController` currently manages the `.secondary` column by itself, but I think if it delegates it to a presenter – similar to how the main sidebar works – it would be better.

```swift
    private func showSecondary(_ viewController: UIViewController) {
        let navigationVC = UINavigationController(rootViewController: viewController)
        splitViewController?.setViewController(navigationVC, for: .secondary)
    }
```

kean commented on WordPress/Classes/System/SplitViewRootPresenter+Notifications.swift:
```diff
@@ -0,0 +1,44 @@
+        notificationsViewController.isSidebarModeEnabled = true
+    }
+
+    func displayed(in splitVC: UISplitViewController) {
```
Is there a need for both `displayed(in:)` and `refresh(with:)`? If you tap on the existing item, it should be a no-op. If you select a new item, it should just display it.

crazytonyli replied:
If we don't use this `refresh(with:)` function, there is a bug in going back to the selected site where the secondary column is an empty UINavigationController.

The root cause is, when displaying the selected site, the secondary UINavigationController is not displayed until some code in `SiteMenuListViewController` runs.

Also, in the current design, the supplementary column is not mutable, but in theory the secondary column can be changed to any `UINavigationController`. So, I think it's fail-safe mechanism to manually sync the `SplitViewDisplayable` properties with the actual `UISplitViewController` content.

kean commented on WordPress/Classes/System/SplitViewRootPresenter.swift:
```diff
@@ -353,6 +367,20 @@ final class SplitViewRootPresenter: RootViewPresenter {
         }
     }
+
+    private func handleLowMemoryWarning() {
```
I would suggest removing it as it complicates the presenter without clear benefits. The view controllers themselves don't take up a lot of space in memory. The only stuff that does is media, and the app clears caches when needed.

kean commented:
It’s a great start. I didn’t think of preserving the state when switching tabs, but it makes so much sense. In terms of the architecture, it’s a great idea to extract the individual tabs.

I noticed an issue where when you switch from “Notifications” to “Reader”, the main sidebar loses the selection state or shows the previous selection, but I think I also saw this in trunk.

crazytonyli committed:
Merge branch 'trunk' into extract-split-view-content-in-dedicated-types

crazytonyli committed:
Add WelcomeSplitViewContent

crazytonyli committed:
Move updating UI code into the `configure(selection:)` function

crazytonyli committed:
Rename `SplitViewDisplayable` properties

crazytonyli committed:
Remove `SplitViewDisplayable.selection`

crazytonyli commented:
@kean I believe I have addressed all your comment. Do you mind having another look?

> I noticed an issue where when you switch from “Notifications” to “Reader”, the main sidebar loses the selection state or shows the previous selection

I couldn't reproduce this issue...

kean commented:
The updates are looking good!

I fixed the menu showing incorrect selection issue separately here https://github.com/wordpress-mobile/WordPress-iOS/pull/23595. 

In term of memory warnings, I think it's fine to keep, but I'm not sure there is much value. I'd rather remove them.

crazytonyli committed:
Remove handling low memory warning

crazytonyli commented:
> In term of memory warnings, I think it's fine to keep, but I'm not sure there is much value. I'd rather remove them.

Ops. I forgot about your comment about the memory warnings. I'm okay with removing that.

I didn't really measure how much memory the Notifications and Reader may take. It's just a gut feeling that they _may_ take up a not small amount of memory because there are potentially a few view controllers in those "tabs", not just the one at the root level.

crazytonyli committed:
Merge branch 'trunk' into extract-split-view-content-in-dedicated-types
